apprise/plugins/NotifyMatrix.py

```diff
@@ -0,0 +1,253 @@
+# matrix webhooks bridge at
+# https://github.com/turt2live/matrix-appservice-webhooks. You'll need
+# to follow the instructions to create a new webhook. You will receive a
+# URL and you can then use that URL to configure this service.
+# See the wiki for more details.
+
+import re
+import requests
+from json import dumps
+from time import time
+
+from .NotifyBase import NotifyBase
+from .NotifyBase import HTTP_ERROR_MAP
+from ..common import NotifyImageSize
+from ..utils import compat_is_basestring
+
+# Token required as part of the API request
+VALIDATE_TOKEN = re.compile(r'[A-Za-z0-9]{64}')
+
+# Default User
+MATRIX_DEFAULT_USER = 'apprise'
+
+# Extend HTTP Error Messages
+MATRIX_HTTP_ERROR_MAP = HTTP_ERROR_MAP.copy()
+MATRIX_HTTP_ERROR_MAP.update({
+    403: 'Unauthorized - Invalid Token.',
+})
+
+class MatrixNotificationMode(object):
+    SLACK = 0
+    MATRIX = 1
+
+MATRIX_NOTIFICATION_MODES = (
+    MatrixNotificationMode.SLACK,
+    MatrixNotificationMode.MATRIX,
+)
+
+class NotifyMatrix(NotifyBase):
+    """
+    A wrapper for Matrix Notifications
+    """
+
+    # The default descriptive name associated with the Notification
+    service_name = 'Matrix'
+
+    # The services URL
+    service_url = 'https://matrix.org/'
+
+    # The default protocol
+    protocol = 'matrix'
+
+    # The default secure protocol
+    secure_protocol = 'matrixs'
+
+    # A URL that takes you to the setup/help of the specific protocol
+    setup_url = 'https://github.com/caronc/apprise/wiki/Notify_matrix'
+
+    # The maximum allowable characters allowed in the body per message
+    body_maxlen = 1000
+
+    def __init__(self, token, mode=None, **kwargs):
+        """
+        Initialize Matrix Object
+        """
+        super(NotifyMatrix, self).__init__(**kwargs)
+
+        if self.secure:
+            self.schema = 'https'
+
+        else:
+            self.schema = 'http'
+
+        if not isinstance(self.port, int):
+            self.notify_url = '%s://%s/api/v1/matrix/hook' % (self.schema, self.host)
+
+        else:
+            self.notify_url = '%s://%s:%d/api/v1/matrix/hook' % (self.schema, self.host, self.port)
+
+        if not VALIDATE_TOKEN.match(token.strip()):
+            self.logger.warning(
+                'The API token specified (%s) is invalid.' % token,
+            )
+            raise TypeError(
+                'The API token specified (%s) is invalid.' % token,
+            )
+
+        # The token associated with the webhook
+        self.token = token.strip()
+
+        if not self.user:
+            self.logger.warning(
+                'No user was specified; using %s.' % MATRIX_DEFAULT_USER)
+            self.user = MATRIX_DEFAULT_USER
+
+        if not mode:
+            self.logger.warning(
+                'No mode was specified, using Slack mode')
+            self.mode = MatrixNotificationMode.SLACK
+
+        else:
+            self.mode = mode
+
+        self._re_formatting_map = {
```

Comment:
This code (with respect to the formatting map) is similar to the static function ```NotifyBase.escape_html()```.  I wonder if there is a way to migrate this into that function (thus improving it) or if it's just better here (since **\r\n** changes to **\\n** instead of **<br/>**)? At this time the _Telegram plugin_ is the only other service using it.

Reply:
I don't know, Slack's documentation specifically says not to HTML-escape everything, but I think the webhook service for Matrix doesn't care too much if you just escape everything. I will have to check that though.